protocols/randhound/handlers.go

```diff
@@ -0,0 +1,375 @@
+		Responses: responses,
+	}
+	return rh.SendTo(rh.Parent(), rh.Peer.r3)
+}
+
+func (rh *RandHound) handleR3(r3 WR3) error {
+
+	// If we are not the root, forward r3 to parent
+	if !rh.Node.IsRoot() {
+		if err := rh.SendTo(rh.Parent(), &r3.R3); err != nil {
+			return err
+		}
+	} else {
+
+		// Verify reply
+		if !bytes.Equal(r3.R3.HI3, rh.hash(rh.SID, rh.hash(rh.SID, rh.Leader.i2.Rc))) {
+			return errors.New(fmt.Sprintf("R3: peer %d replied to wrong I3 message", r3.Src))
+		}
+
+		// Collect replies of the peers
+		rh.Leader.r3[r3.Src] = &r3.R3
+
+		invalid := []uint32{}
+		for _, r3resp := range rh.Leader.r3[r3.Src].Responses {
+
+			resp := &poly.Response{}
+			resp.UnmarshalInit(rh.Node.Suite())
+			if err := resp.UnmarshalBinary(r3resp.Resp); err != nil {
+				return err
+			}
+
+			// Verify that response is securely bound to promise and mark invalid ones
+			if err := rh.Leader.states[r3resp.DealerIdx].AddResponse(int(r3resp.ShareIdx), resp); err != nil {
+				invalid = append(invalid, r3resp.DealerIdx)
+			}
+		}
+		rh.Leader.invalid[r3.Src] = &invalid
+
+		// Continue, once all replies have arrived
+		if uint32(len(rh.Leader.r3)) == rh.Group.N-1 {
+			rh.Leader.i4 = &I4{
+				SID:     rh.SID,
+				Invalid: rh.Leader.invalid,
+			}
+			return rh.sendToChildren(rh.Leader.i4)
+		}
+	}
+	return nil
+}
+
+func (rh *RandHound) handleI4(i4 WI4) error {
+
+	// If we are not a leaf, forward i4 to children
+	if !rh.Node.IsLeaf() {
+		if err := rh.sendToChildren(&i4.I4); err != nil {
+			return err
+		}
+	}
+
+	// Verify session ID
+	if !bytes.Equal(rh.SID, i4.I4.SID) {
+		return errors.New(fmt.Sprintf("I4: peer %d received message with incorrect session ID", rh.nodeIdx()))
+	}
+
+	rh.Peer.i4 = &i4.I4
+
+	// Remove all invalid shares and prepare hash buffer
+	buf := new(bytes.Buffer)
+	invalid := rh.Peer.i4.Invalid[rh.nodeIdx()]
+	for _, dealerIdx := range *invalid {
+		delete(rh.Peer.shares, dealerIdx)
+		if err := binary.Write(buf, binary.LittleEndian, dealerIdx); err != nil {
+			return err
+		}
+	}
+
+	rh.Peer.r4 = &R4{
+		Src: rh.nodeIdx(),
+		HI4: rh.hash(
+			rh.SID,
+			buf.Bytes()),
+		Shares: rh.Peer.shares,
+	}
+	return rh.SendTo(rh.Parent(), rh.Peer.r4)
+}
+
+func (rh *RandHound) handleR4(r4 WR4) error {
+
+	// If we are not the root, forward r4 to parent
+	if !rh.Node.IsRoot() {
+		if err := rh.SendTo(rh.Parent(), &r4.R4); err != nil {
+			return err
+		}
+	} else {
+
+		// Verify reply
+		buf := new(bytes.Buffer)
+		invalid := rh.Leader.i4.Invalid[r4.Src]
+		for _, dealerIdx := range *invalid {
+			if err := binary.Write(buf, binary.LittleEndian, dealerIdx); err != nil {
+				return err
+			}
+		}
+		if !bytes.Equal(r4.R4.HI4, rh.hash(rh.SID, buf.Bytes())) {
+			return errors.New(fmt.Sprintf("R4: peer %d replied to wrong I4 message", r4.Src))
+		}
+
+		// Collect replies of the peers
+		rh.Leader.r4[r4.Src] = &r4.R4
+
+		// Continue, once all replies have arrived
+		if uint32(len(rh.Leader.r4)) == rh.Group.N-1 {
```

Comment:
If the comment talks about the branch as a whole, it's before the branch (like `Verify share` below). Here it is about what if the branch is taken, so it's after the `if`.